# 🐼 Pandas - Class 8: Working with Time Series
Welcome to **Class 8** of our Pandas series. Today we’ll learn how to handle and analyze time series data.

## 1. Converting Columns to Datetime (`pd.to_datetime`)
- Use `pd.to_datetime()` to convert strings or numbers into datetime objects.
- Once converted, Pandas recognizes them as dates, enabling powerful operations.

In [ ]:
import pandas as pd

# Sample data for the DataFrame
data = {
    'Order_ID': [101, 102, 103, 104, 105],
    'Order_Date': [
        '2023-01-15',      # YYYY-MM-DD
        '02/28/2023',      # MM/DD/YYYY
        '10-Mar-23',       # DD-Mon-YY
        '20230401',        # YYYYMMDD
        '2023-05-20'
    ],
    'Customer': ['Alice', 'Bob', 'Charlie', 'David', 'Eve'],
    'Account': ['A101', 'B202', 'C303', 'D404', 'E505']
}

# Create the DataFrame
df = pd.DataFrame(data)

# Display the DataFrame
print(df)
print("\nDataFrame Info:")
print(df.info())

   Order_ID  Order_Date Customer Account
0       101  2023-01-15    Alice    A101
1       102  02/28/2023      Bob    B202
2       103   10-Mar-23  Charlie    C303
3       104    20230401    David    D404
4       105  2023-05-20      Eve    E505

DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Order_ID    5 non-null      int64 
 1   Order_Date  5 non-null      object
 2   Customer    5 non-null      object
 3   Account     5 non-null      object
dtypes: int64(1), object(3)
memory usage: 292.0+ bytes
None


In [ ]:
# errors = "coerce" forcefully solves errors

df["Order_Date"] = pd.to_datetime(df['Order_Date'], errors = 'coerce')
display(df)

,Order_ID,Order_Date,Customer,Account
0,101,2023-01-15,Alice,A101
1,102,NaT,Bob,B202
2,103,NaT,Charlie,C303
3,104,NaT,David,D404
4,105,2023-05-20,Eve,E505


## 2. Setting DateTime Index & Resampling
- Set a datetime column as the index with `set_index()`.
- Resample time series using `resample('M').mean()` or other rules ('D', 'W', 'Y').
- Useful for aggregating data by time periods.

In [ ]:
df_idx = df.set_index("Order_Date")
display(df_idx)

,Order_ID,Customer,Account
Order_Date,,,
2023-01-15,101,Alice,A101
NaT,102,Bob,B202
NaT,103,Charlie,C303
NaT,104,David,D404
2023-05-20,105,Eve,E505


In [ ]:
df_idx.resample('M')["Account"].sum()

/tmp/ipykernel_220/2551877499.py:1: FutureWarning: 'M' is deprecated and will be removed in a future version, please use 'ME' instead.
  df_idx.resample('M')["Account"].sum()


,Account
Order_Date,
2023-01-31,A101
2023-02-28,0
2023-03-31,0
2023-04-30,0
2023-05-31,E505


## mini-project

Converting columns to datetime (pd.to_datetime)

Setting a DateTime index & resampling

In [ ]:

# 1️⃣ Create a dataset of website visits (string dates + visit counts)

# 2️⃣ Convert VisitDate to datetime
# Write your code here

# 3️⃣ Set VisitDate as the index
# Write your code here

# 4️⃣ Resample to get total visitors per month
# Write your code here

# 5️⃣ Resample to get average visitors per week
# Write your code here


---
## Summary
- Converted columns to datetime with `pd.to_datetime`.
- Set DateTime index and resampled data.
- Performed date-based filtering and slicing.
- Applied rolling and expanding window calculations.

# Task
Create a pandas DataFrame representing website visits with `VisitDate` (strings in various formats) and `VisitCount` columns, convert `VisitDate` to datetime objects, set `VisitDate` as the DataFrame's index, then resample the data to calculate the total visitors per month and the average visitors per week. Finally, summarize the insights from the resampled data.

## Create Website Visits Dataset

### Subtask:
Create a pandas DataFrame with 'VisitDate' (as strings in various formats) and 'VisitCount' columns to represent website visits. This will be the initial dataset for the mini-project.


**Reasoning**:
I will create a pandas DataFrame named `website_df` from a dictionary containing 'VisitDate' and 'VisitCount' columns as specified in the instructions to initialize the dataset for the mini-project.



In [ ]:
import pandas as pd

# 1. Define a dictionary named website_data
website_data = {
    'VisitDate': [
        '2023-01-01',      # YYYY-MM-DD
        '01/08/2023',      # MM/DD/YYYY
        'Jan 15, 2023',    # Mon DD, YYYY
        '20230122',        # YYYYMMDD
        '2023-01-29',      # YYYY-MM-DD
        '02-05-2023',      # MM-DD-YYYY
        'Feb 12, 2023',    # Mon DD, YYYY
        '20230219',        # YYYYMMDD
        '2023-02-26'       # YYYY-MM-DD
    ],
    'VisitCount': [120, 150, 130, 180, 110, 160, 145, 190, 125]
}

# 4. Create a pandas DataFrame named website_df from this dictionary
website_df = pd.DataFrame(website_data)

# Display the DataFrame to verify
print("Website Visits DataFrame:")
print(website_df)
print("\nDataFrame Info:")
print(website_df.info())

Website Visits DataFrame:
      VisitDate  VisitCount
0    2023-01-01         120
1    01/08/2023         150
2  Jan 15, 2023         130
3      20230122         180
4    2023-01-29         110
5    02-05-2023         160
6  Feb 12, 2023         145
7      20230219         190
8    2023-02-26         125

DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   VisitDate   9 non-null      object
 1   VisitCount  9 non-null      int64 
dtypes: int64(1), object(1)
memory usage: 276.0+ bytes
None


**Reasoning**:
The next logical step, as indicated by the task description and the notebook's structure, is to convert the 'VisitDate' column in `website_df` from string objects to proper datetime objects using `pd.to_datetime()`. This will enable time-series operations.



In [ ]:
website_df['VisitDate'] = pd.to_datetime(website_df['VisitDate'])

print("DataFrame after converting 'VisitDate' to datetime:")
print(website_df.info())

ValueError: time data "01/08/2023" doesn't match format "%Y-%m-%d", at position 1. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.

**Reasoning**:
The previous code failed because the 'VisitDate' column contains dates in various formats. To correctly parse these mixed formats, I need to use the `format='mixed'` argument within `pd.to_datetime()`.



In [ ]:
website_df['VisitDate'] = pd.to_datetime(website_df['VisitDate'], format='mixed')

print("DataFrame after converting 'VisitDate' to datetime:")
print(website_df.info())

DataFrame after converting 'VisitDate' to datetime:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   VisitDate   9 non-null      datetime64[ns]
 1   VisitCount  9 non-null      int64         
dtypes: datetime64[ns](1), int64(1)
memory usage: 276.0 bytes
None


**Reasoning**:
Now that the 'VisitDate' column is in the correct datetime format, I will set it as the DataFrame's index using `set_index()` to prepare for time series operations like resampling.



In [ ]:
website_df = website_df.set_index('VisitDate')

print("DataFrame after setting 'VisitDate' as index:")
print(website_df.head())
print("\nDataFrame Info:")
print(website_df.info())

DataFrame after setting 'VisitDate' as index:
            VisitCount
VisitDate             
2023-01-01         120
2023-01-08         150
2023-01-15         130
2023-01-22         180
2023-01-29         110

DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 9 entries, 2023-01-01 to 2023-02-26
Data columns (total 1 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   VisitCount  9 non-null      int64
dtypes: int64(1)
memory usage: 144.0 bytes
None


**Reasoning**:
Now that 'VisitDate' is the datetime index, I will resample the `website_df` to calculate the total visitors per month and the average visitors per week, as required by the task.



In [ ]:
monthly_total_visitors = website_df['VisitCount'].resample('ME').sum()
weekly_average_visitors = website_df['VisitCount'].resample('W').mean()

print("\nTotal Visitors Per Month:")
print(monthly_total_visitors)

print("\nAverage Visitors Per Week:")
print(weekly_average_visitors)


Total Visitors Per Month:
VisitDate
2023-01-31    690
2023-02-28    620
Freq: ME, Name: VisitCount, dtype: int64

Average Visitors Per Week:
VisitDate
2023-01-01    120.0
2023-01-08    150.0
2023-01-15    130.0
2023-01-22    180.0
2023-01-29    110.0
2023-02-05    160.0
2023-02-12    145.0
2023-02-19    190.0
2023-02-26    125.0
Freq: W-SUN, Name: VisitCount, dtype: float64


### Summary of Resampled Data Insights

From the resampled data, we can observe the following insights:

1.  **Monthly Total Visitors (`monthly_total_visitors`):**
    *   **January 2023 (2023-01-31):** The website received a total of 690 visitors.
    *   **February 2023 (2023-02-28):** The total visitors slightly decreased to 620. This could indicate a slight dip in engagement or traffic during February.

2.  **Average Weekly Visitors (`weekly_average_visitors`):**
    *   The average weekly visitors show fluctuations throughout January and February.
    *   **Peak Week:** The week ending 2023-02-19 had the highest average visitors with 190, suggesting a particular event or content might have driven traffic that week.
    *   **Lowest Week:** The week ending 2023-01-29 had the lowest average visitors with 110.
    *   The data provides a granular view of visitor trends, which could be useful for identifying popular periods or weeks that need attention.

## Summary:

### Data Analysis Key Findings

*   The initial `VisitDate` column contained various string formats, requiring the use of `pd.to_datetime()` with `format='mixed'` for successful conversion to datetime objects.
*   After converting `VisitDate` to datetime objects, the DataFrame's index was successfully set to `VisitDate`, enabling time-series analysis.
*   **Monthly Total Visitors**:
    *   January 2023 recorded a total of 690 visitors.
    *   February 2023 showed a slight decrease, with a total of 620 visitors.
*   **Average Weekly Visitors**:
    *   The week ending 2023-02-19 had the highest average visitors at 190.
    *   The week ending 2023-01-29 had the lowest average visitors at 110.

### Insights or Next Steps

*   The observed slight decrease in total visitors from January to February could warrant further investigation into marketing efforts or content performance during these months.
*   Analyzing the specific events or content released around the peak week ending 2023-02-19 could help identify successful strategies that drove higher visitor engagement.
